READMES/environments.md
# Environments & the Content Build Process

To enable end to end quality analysis, from the VA.gov CMS through to the public facing website, new feature review, and stakeholder demonstrations, multiple joined CMS/WEB environments may be created by any user with access to Tugboat. (More info on [Tugboat](https://va.ghe.com/software/va.gov-cms/blob/main/READMES/tugboat.md).)

The following table lists all environments and CMS/WEB sites used in the development process for VA.gov:

| Environment | Drupal (CMS) | Frontend (FE / WEB / Static) | FE Data source | Management |
| -- | -- | -- | -- | -- |
| **PROD**<br> Live Site | [prod.cms.va.gov](https://prod.cms.va.gov)<br/>va.gov-cms: main | [www.va.gov](https://www.va.gov)<br/>content-build: main<br/>vets-website: main | prod.cms.va.gov | [BRD: Jenkins](http://jenkins.vfs.va.gov/job/deploys/view/Prod/job/cms-vagov-prod/) |
| **STAGING** <br> Pre-release testing. | [staging.cms.va.gov](https://staging.cms.va.gov)<br/>va.gov-cms: main<br/><br/>Staging CMS uses Tugboat Preview [content-build-branch-build (CMS PROD Mirror)](https://tugboat.vfs.va.gov/6189a9af690c68dad4877ea5) as its data source. This Preview is refreshed daily at 7am UTC (2am EST, 1am EDT), using the latest Database and Asset file snapshot from AWS S3 at that time.<br/><br/> * Staging CMS database edits will be overwritten the next time code is merged.<br/>* Staging CMS is not the source of data for the Staging front-end, meaning: edits you make in the Staging CMS will not cause the Staging front-end to be re-built, and your changes will not appear in the Staging.va.gov front-end. | [staging.va.gov](http://staging.va.gov)<br/>content-build: main<br/>vets-website: main | Tugboat Preview [content-build-branch-build (CMS PROD Mirror)](https://tugboat.vfs.va.gov/6189a9af690c68dad4877ea5) refreshed daily at 7am UTC (2am EST, 1am EDT), using the latest Database and Asset file snapshot from AWS S3 at that time.  | [BRD: Jenkins](http://jenkins.vfs.va.gov/job/deploys/view/Staging/job/cms-vagov-staging/) |
| [LOCAL](http://va-gov-cms.ddev.site) <br> Local development | [va-gov-cms.ddev.site](http://va-gov-cms.ddev.site) | [va-gov-cms.ddev.site](http://va-gov-cms.ddev.site) <br> [va-gov-cms.ddev.site/static](http://va-gov-cms.ddev.site) <br> [va-gov-cms.ddev.site/$URL?\_format=static_html](http://va-gov-cms.ddev.site/$URL) | |
| [CI / PR](1) <br> Pull Requests & Automated Testing | pr###-{hash}.ci.cms.va.gov<br/><br/>va.gov-cms: specified branch | web-{hash}.ci.cms.va.gov <br> pr###-{hash}.ci.cms.va.gov/static <br> pr###-{hash}.ci.cms.va.gov/$URL?\_format=static_html<br/><br/>content-build: main, unless another branch is specified in the CMS<br/>vets-website: main<br/><br/>It is not possible to specify an alternate vets-website branch until [#6434](https://va.ghe.com/software/va.gov-cms/issues/6434) is complete | Tugboat: CMS > [CMS Pull Request Environments](https://tugboat.vfs.va.gov/5fd3b8ee7b4657022b5722d6#RepoBasePreviews__heading-h2--3rTIH) > Base Preview, including Database and Asset files, refreshed daily at 10am UTC (5am EST, 4am EDT).<br/><br/>Tugboat PR Previews need to be rebuilt (after 5am EST), or refreshed, for the database to stay up-to-date with Prod. | [CMS-CI: Tugboat](1) |
| [Demos](2) <br> Demos & Training | cms-{hash}.demo.cms.va.gov | web-{hash}.demo.ci.cms.va.gov <br> cms-{hash}.demo.cms.va.gov/static <br> cms-{hash}.demo.cms.va.gov/$URL?\_format=static_html | Tugboat: CMS > [CMS Demo Environments](https://tugboat.vfs.va.gov/5ffe2f4dfa1ca136135134f6) > Base Preview, including Database and Asset files, refreshed daily at 9am UTC (4am EST, 3am EDT).<br/><br/>Demo Previews need to be rebuilt (after 4am EST), or refreshed, for the database to stay up-to-date with Prod. | [CMS-CI: Tugboat](1) |

### Review Instances
**NOTE**: [Review Instances](https://depo-platform-documentation.scrollhelp.site/developer-docs/review-instances) are another system that provides test environments, for CI PRs in vets-website and vets-api repos. RIs are managed by the Platform team. (Feb 2023: Platform has an active project underway to replace Review Instances; Ray Messina, POC.) That's not a CMS concern, other than to note: Tugboat does also indirectly act as the Content Data Source for RIs.
| Environment | Drupal (CMS) | Frontend (FE / WEB / Static) | FE Data source | Management |
| -- | -- | -- | -- | -- |
| CI / PR for vets-website & vets-api<br> Pull Requests & Automated Testing | None provided | {hash}.review.vetsgov-internal<br/><br/>* vets-website or vets-api: specified branch<br/>* content-build: main<br/>| Review instances can't pull data directly from Tugboat. To work around this, when content-build merges to main, a process captures the data + files from CI (effectively, Tugboat Preview [content-build-branch-build (CMS PROD Mirror)](https://tugboat.vfs.va.gov/6189a9af690c68dad4877ea5) 7am UTC or latest refreshed data/assets) and pushes them to S3 to a different location than the every 15mins data + asset backup. Review Instances then pull and use the new S3 snapshot archive file. Therefore: the freshness of the data on a Review Instance will line up with timing for the last time content-build had a merge to main.  | NA |

### Using Tugboat to preview Next Build and Content Build changes

Tugboat can be used to preview changes being made in Next Build and Content Build. This can be useful if you need to check Next Build or Content Build changes together with new CMS code, or if you want to create custom content to test features of the frontend code that VA.gov content doesn't test or demonstrate well.

To create your Tugboat instance, do one of the following:

1. When **testing changed CMS code**, simply create a PR with your working branch. If your CMS code changes are not yet ready for review, create a draft PR. This will still create a Tugboat instance.
2. When **testing frontend changes with no CMS code changes**, please create a Demo CMS Tugboat instance and do your frontend testing there. You should delete this demo instance once you are finished testing.

#### Testing Next Build
Next Build is available on CMS Tugboat at the Tugboat domain with `next-` in it, for example https://next-medc0xjkxm4jmpzxl3tfbcs7qcddsivh.ci.cms.va.gov/. By default this uses the default next-build and vets-website branches.

This can be changed via the form at `/admin/content/deploy/next`. At that URL, you will see a simple form that allows you to rebuild the frontend with the option of choosing a different branch for either next-build or vets-website. Choosing 'Select a different next-build branch/pull request' will reveal a text field. Entering a search term in that text field will reveal both branches and PRs that match that search term.

![The form for changing Next Build branches, with the search text "update" entered and results showing.](images/tugboat-next-build-form.png)

Once you've chosen the branches you want to test with, hit 'Restart Next Build Server'. You will see status information about the build at the bottom of the page. You may need to refresh the page for this information to fill in. The Log File will be the best way to observe the progress of the build.

![Status Details of the Next Build Server Rebuild process.](images/tugboat-next-build-status.png)

Rebuilding the Next Build Server generally takes about 5 minutes. Once complete, you can view the restarted server at the standard Next Build URL.

#### Testing Content Build
Next Build is available on CMS Tugboat at the Tugboat domain with `web-` in it, for example https://web-medc0xjkxm4jmpzxl3tfbcs7qcddsivh.ci.cms.va.gov/. By default this uses the default content-build and vets-website branches.

This can be changed via the form at `/admin/content/deploy`. At that URL, you will see a simple form that allows you to rebuild the frontend with the option of choosing a different branch for either content-build or vets-website. Choosing 'Select a different content-build branch/pull request' will reveal a text field. Entering a search term in that text field will reveal both branches and PRs that match that search term, similarly to the Next Build form described above.

Once you've chosen the branches you want to test with, hit 'Release Content'. You will see status information about the build at the bottom of the page. You may need to refresh the page for this information to fill in. The Log File will be the best way to observe the progress of the build.

Rebuilding the Content Build static content generally takes about 60 minutes. Once complete, you can view the restarted server at the standard Content Build URL.

### Other
* [Access](./access.md): how to access these environments

## What is an Environment?

_Environments_ are copies of the production site that are running newer
code or have different content that needs to be tested before going live.

_Environments_ can also be used for demonstrations or training, without worrying
about disrupting production content.

Each _Environment_ has both a _CMS_ and a _WEB_ site. The _CMS_ is a content management
system built with Drupal, and the _WEB_ site is a static HTML site, built with Metalsmith.

The _WEB_ build process consumes the content from the _CMS_ in the same environment.


## Important Concepts

- Each _WEB_ site is made up of generated "static" files. This means that the _WEB_ site reflects the content from the CMS _at the time the WEB Build process was run_.
- _CMS_ Editors will not see changes in the _WEB_ site until a "WEB Build" is triggered and the process completes successfully.
- The _WEB_ site will not be accessible until at least one "WEB Build" has run successfully. This happens automatically for CI environments, but not yet for Demo environments. If you get an error such as "Forbidden" when visiting the _WEB_ site, try running the _Rebuild WEB_ process again.


## Manually build a demo env: Step-by-step Instructions

1. Visit [https://tugboat.vfs.va.gov/](https://tugboat.vfs.va.gov/) and click on the "GitHub" link to log in with GitHub.

   ![CMS-CI Homepage](images/tugboat-home.png)

2. Click the CMS link and the click the "CMS Demo Environments" to visit the Demos page.

   ![Demo environments page](images/tugboat-demos.png)

3. Scroll down to the "Base Previews" heading and select "Clone" from the "Actions" menu.

   ![Clone Preview](images/tugboat-clone-preview.png)

4. Your demo environment has been created. It will be titled "master". Scroll to the top of the page, click on the "Settings" link for your new environment, and give it a title starting with your initials and a hyphen. Then, click the "Save Configuration" button.

   ![Preview Settings](images/tugboat-preview-settings.png)

5. If you wish to copy [prod.cms.va.gov](http://prod.cms.va.gov) using the latest content, do not change any additional options.

6. Click the "Preview" button for your environment to visit it.

   ![Environment Preview](images/tugboat-preview.png)

**NOTE:** The WEB site for this environment will not work until you trigger a WEB Build process on the "Release Content" page.

### WEB Build Process

Within each environment, the static HTML for the _WEB_ site is occasionally
"rebuilt" so that the latest content from that environment's _CMS_ is used.

The _WEB_ build process is tested in the CI system to ensure compatibility with the
CMS content schema.

The _WEB_ build process is triggered automatically by certain actions in the CMS
or manually via the [command line](#cli-build).

#### Build Triggers

The _WEB_ instance of an environment is rebuilt when any of the following actions take place in the _CMS_:

- Facility Alert or Individual Facility Operating Status is created or updated.
- The "Rebuild WEB" button is pressed.
- @TODO: Document all current build triggers.

_Note to Developers:_ Keep this list up to date to help content editors understand the process.

## Rebuilding Environments Manually

There is a special button and form for rebuilding VA.gov environments. Use this
to manually trigger either a WEB or CMS rebuild (or both), and optionally check
out different code.

#### Step-by-step Instructions

1. Find the environment you would like to build in the [CMS-CI](1) site, and go to the preview.
2. Hover over the "Content" admin menu, and then click "Release Content".

   ![Environment Rebuild](images/tugboat-rebuild.png)

3. If you wish to just trigger a WEB rebuild with the existing content, do not change any other options.

   ![Release Content Form](images/tugboat-release-content.png)

   **For advanced users and developers:**

   1. If you wish to change the branch of the WEB project, click "Select a different frontend branch/pull request". Start typing to search by the name of the branch or pull request you wish to use.

   ![Release Content Web Branch](images/tugboat-release-content-web-branch.png)

4. Press the "Release content" button. You will see the status of your build under the "Wait for the release to complete" heading. You may view the build logs by clicking the "View logs" link.

   ![Content Release Status](images/tugboat-content-release-status.png)

5. Once the REBUILD process is complete, you can click the "Go to front end" button under the "Access the frontend environment" heading.

   ![WEB Link](images/tugboat-web-link.png)

6. That's it! If the process completed, you should see a site that looks like VA.gov.



## Creating new Environments

Within the [CMS-CI](1) platform, in "CMS Demo Environments", users can create new environments by cloning the "master" Base Preview.

## Hosting Architecture

All environments are hosted on VA GovCloud in AWS GovCloud.

The primary environments, DEV, STAGING, and PROD, are hosted in the _BRD_ system.

Pull Request Environments and Ad Hoc environments are hosted in the _CMS-CI_ system.

SOCKS proxy or PIV+GFE hardware is required for accessing VA internal network.

### BRD: Jenkins

> Build, Release, Deploy

[jenkins.vfs.va.gov](http://jenkins.vfs.va.gov/)

- Source Code: https://va.ghe.com/software/devops/tree/master/ansible/build/roles/cms
- Runs Continuous Integration for about a dozen different applications with different requirements, including `vets-website`, `vets-api`, `cms`, and soon `cms-ci`.
- BRD Process is standardized across apps using Ansible playbooks and roles. See the [DevOps Repo Documentation](https://va.ghe.com/software/devops/blob/master/README.md) for more information.

  - The "Build" process creates the entire server image and permantently tags and archives it as an AMI.

CHANGELOG.md

# Changelog
All notable changes to this project will be documented in this file.

The format is based on [git-release-notes](https://github.com/ariatemplates/git-release-notes).


## [Unreleased]

* __Nothing unreleased__

## 2019-09-24.1 Sprint 23 release 1. tag  20190924.1
*  __VAGOV-6018: disable devel and remove test (#590)__
    2019-09-23 09:28:48 -0400 - (Steve Wirt) [5753c272]
*  __VAGOV-6017: Replace legacy support email with new. (#592)__
    2019-09-20 17:21:29 -0400 - (Ethan Teague) [c574c314]
*  __VAGOV-5935: Adding new image style and flag. (#578)__
    2019-09-19 13:44:40 -0400 - (Ethan Teague) [14a016ee]
*  __VAGOV-2214 Add SSOi documentation. (#579)__
    2019-09-19 13:28:01 -0400 - (Steve Wirt) [1e91e351]
*  __VAGOV-5835: Remove unused legacy flags. (#585)__
    2019-09-19 13:25:27 -0400 - (Ethan Teague) [8073e70d]
*  __VAGOV-5835 Update vets-website to latest. (#587)__
    2019-09-19 12:56:09 -0400 - (Elijah Lynn) [9d56fda4]
*  __VAGOV-5984 Update header menu promo reference (#586)__
    2019-09-19 12:43:09 -0400 - (jonbot) [41c59db7]
*  __VAGOV-5809: Updating tablefield to prevent log notices. (#588)__
    2019-09-19 10:18:02 -0400 - (Ethan Teague) [e00a83a5]
*  __Update https.md README with OS X tips for importing VA-Internal certs. (#584)__
    2019-09-18 16:31:08 -0400 - (Kevin Walsh) [68fdb4f4]
*  __VAOGV-000 Move fast tests to beginning for faster feedback. (#583)__
    2019-09-18 16:27:54 -0400 - (Elijah Lynn) [d1034366]
*  __VAGOV-0000 SSOi fix (#582)__
    2019-09-17 20:24:04 -0400 - (Steve Wirt) [0f82552c]
*  __VAGOV-5892 Adjust SSO settings. (#565)__
    2019-09-17 10:04:24 -0700 - (Steve Wirt) [c124a67a]
*  __VAGOV-2303: README.md (#567)__
    2019-09-13 11:12:24 -0400 - (Jon Pugh) [1d035143]
*  __VAGOV-5913: Adding cac role test for menu access. (#577)__
    2019-09-12 11:43:08 -0400 - (Steve Wirt) [9ba5dd35]
*  __VAGOV-5766: Changing node type meta tag order. (#575)__
    2019-09-11 16:22:02 -0400 - (Ethan Teague) [5e3492f9]

## 2019-09-10.1 Sprint 22 release 2. tag  20190910.1
*  __VAGOV-000 Merge CHANGELOG.md from master.__
    2019-09-10 17:44:47 -0400 - (Steve Wirt) [03d5064e]
*  __VAGOV-5776: Move top tasks menu ownership to Global UX team. (#574)__
    2019-09-10 16:11:20 -0400 - (Kevin Walsh) [9d886fe2]
*  __VAGOV-5908: Config ignore menu breadcrumbs settings. (#566)__
    2019-09-10 15:42:30 -0400 - (Kevin Walsh) [3cd3b8eb]
*  __VAGOV-5776: Add content admin role. (#572)__
    2019-09-10 14:19:25 -0400 - (Kevin Walsh) [ca460e48]
*  __VAGOV-5857: Adding embedded asset functionality to wysiwyg. (#571)__
    2019-09-10 14:16:17 -0400 - (Ethan Teague) [5c76e5f6]
*  __VAGOV-5865: Bottom Rail Footer Config (#569)__
    2019-09-10 13:16:14 -0400 - (jonbot) [ffa1f5f0]
*  __VAGOV-5884: Content API Consumer Menu Permission (#568)__
    2019-09-10 09:19:36 -0400 - (jonbot) [f41a384e]
*  __VAGOV-5669:Move documentation to github. (#563)__
    2019-09-09 18:07:19 -0400 - (Steve Wirt) [050bcda5]
*  __VAGOV-4871: Header Feature Flag (#570)__
    2019-09-09 17:17:03 -0400 - (jonbot) [a37db289]
*  __VAGOV-000: Empty commit for devops test to trigger CD to DEV. #6__
    2019-09-09 09:37:08 -0700 - (Elijah Lynn) [d425ad77]
*  __VAGOV-5827 Use PROD SSO metadata and endpoint. (#562)__
    2019-09-06 14:53:25 -0400 - (Elijah Lynn) [9ea94e7d]
*  __VAGOV-5890 test commit devops CD to DEV. (#564)__
    2019-09-06 11:44:31 -0700 - (Elijah Lynn) [2c3cf8de]
*  __VAGOV-5890 Test commit to trigger CD to DEV.__
    2019-09-06 11:39:11 -0700 - (Elijah Lynn) [ea628188]
*  __VAGOV-000: Empty commit for devops test to trigger CD to DEV. #4__
    2019-09-05 10:48:49 -0700 - (Elijah Lynn) [759447ff]
*  __VAGOV-000: Empty commit for devops test to trigger CD to DEV. #3__
    2019-09-05 10:47:36 -0700 - (Elijah Lynn) [1ae4635a]
*  __VAGOV-000: Empty commit for devops test to trigger CD to DEV. #2__
    2019-09-05 09:55:39 -0700 - (Elijah Lynn) [4c7357bd]
*  __VAGOV-000: Empty commit for devops test to trigger CD to DEV.__
    2019-09-05 08:45:24 -0700 - (Elijah Lynn) [176e07cb]
*  __VAGOV-5857: Adding all flags to config. (#561)__
    2019-09-04 16:36:54 -0400 - (Ethan Teague) [7359f6b4]
*  __VAGOV-5834: Remove inline entity form widgets. (#560)__
    2019-09-04 09:48:58 -0400 - (Kevin Walsh) [71703908]


## 2019-09-03.2 Sprint 22 release 1. tag  20190903.2

*  __VAGOV-5843: Enabling entity_clone.__
    2019-08-31 16:58:35 -0400 - (Ethan Teague) [61a57b68]
*  __VAGOV-5843: Adding entity_clone to composer.json/composer.lock.__
    2019-08-31 16:56:57 -0400 - (Ethan Teague) [dd058000]
*  __VAGOV-5669: Add build documentation.__
    2019-08-30 23:42:18 -0400 - (Steve Wirt) [4ce28a0c]
*  __VAGOV-5669: Add interfaces documentation.__
    2019-08-30 23:18:53 -0400 - (Steve Wirt) [85db59e3]
*  __VAGOV-5669: Add environment documentation.__
    2019-08-30 17:57:05 -0400 - (Steve Wirt) [2d7fa5eb]
*  __VAGOV-5669: Add conventions documentation.__
    2019-08-30 17:29:47 -0400 - (Steve Wirt) [b9c180de]
*  __VAGOV-5669: Add debug documentation.__
    2019-08-30 17:14:34 -0400 - (Steve Wirt) [764f185d]
*  __VAGOV-5669: Add testing documentation.__
    2019-08-30 17:02:38 -0400 - (Steve Wirt) [c2ba2243]
*  __Update Jenkinsfile__
    2019-08-30 12:42:27 -0600 - (Ryan Watson) [091844ac]
*  __VAGOV-5810 Add global redirect perms to redirect administrator role.__
    2019-08-29 09:42:29 -0400 - (Steve Wirt) [3b207339]
*  __VAGOV-000: Move return outside of for loop.__
    2019-08-28 14:54:56 -0400 - (Beth Jacobson) [94b0e92f]
*  __VAGOV-5807: Fix Owner dropdown on bulk edit view.__
    2019-08-28 09:57:09 -0400 - (Kevin Walsh) [4d214798]
*  __Revert "VAGOV-5553: commit a failing test to confirm devshop behat extension failure output works"__
    2019-08-28 09:14:54 -0400 - (Jon Pugh) [a3ce4e27]
*  __VAGOV-5690: Forcing removal of bangpound/composer-dotenv.__
    2019-08-28 08:56:15 -0400 - (Jon Pugh) [0bb8bff8]

docroot/sites/default/settings.php
<?php

// phpcs:ignoreFile

/**
 * For documentation and more options
 * @see /docroot/sites/default/default.settings.php
 */

/**
 * @file
 * Drupal 8 configuration file.
 */

$databases['default']['default'] = array(
  'driver' => 'mysql',
  'database' => getenv('CMS_MARIADB_DATABASE') ?: 'drupal8',
  'username' => getenv('CMS_MARIADB_USERNAME') ?: 'drupal8',
  'password' => getenv('CMS_MARIADB_PASSWORD') ?: 'drupal8',
  'prefix' => '',
  // 'db' is the default DB container hostname for local.
  'host' => getenv('CMS_MARIADB_HOST') ?: 'db',
  'port' => 3306,
  'namespace' => 'Drupal\\Core\\Database\\Driver\\mysql',
);

/**
 * Salt for one-time login links, cancel links, form tokens, etc.
 *
 * This variable will be set to a random value by the installer. All one-time
 * login links will be invalidated if the value is changed. Note that if your
 * site is deployed on a cluster of web servers, you must ensure that this
 * variable has the same value on each server.
 *
 * For enhanced security, you may set this variable to the contents of a file
 * outside your document root; you should also ensure that this file is not
 * stored with backups of your database.
 *
 * Example:
 * @code
 *   $settings['hash_salt'] = file_get_contents('/home/example/salt.txt');
 * @endcode
 *
 * @todo Change hash_salt and use example above.
 */
$settings['hash_salt'] = 'HJwuroKYGPRzhRHXnWR7H38RkH9rOkJ0WP8C5qD_pRStai8zvFX655aKHZO1gXNXPXhkFRNjoQ';

/**
 * Access control for update.php script.
 *
 * If you are updating your Drupal installation using the update.php script but
 * are not logged in using either an account with the "Administer software
 * updates" permission or the site maintenance account (the account that was
 * created during installation), you will need to modify the access check
 * statement below. Change the FALSE to a TRUE to disable the access check.
 * After finishing the upgrade, be sure to open this file again and change the
 * TRUE back to a FALSE!
 */
$settings['update_free_access'] = FALSE;

/**
 * Load services definition file.
 */
$settings['container_yamls'][] = $app_root . '/' . $site_path . '/../default/services.yml';

/**
 * The default list of directories that will be ignored by Drupal's file API.
 *
 * By default ignore node_modules and bower_components folders to avoid issues
 * with common frontend tools and recursive scanning of directories looking for
 * extensions.
 *
 * @see file_scan_directory()
 * @see \Drupal\Core\Extension\ExtensionDiscovery::scanDirectory()
 */
$settings['file_scan_ignore_directories'] = [
  'node_modules',
  'bower_components',
];

/**
 * The default number of entities to update in a batch process.
 *
 * This is used by update and post-update functions that need to go through and
 * change all the entities on a site, so it is useful to increase this number
 * if your hosting configuration (i.e. RAM allocation, CPU speed) allows for a
 * larger number of entities to be processed in a single batch run.
 */
$settings['entity_update_batch_size'] = 50;

/**
 * Image Style Settings
 *
 * We don't need `itok` DDoS protection in this firewalled environment.
 */
$config['image.settings']['allow_insecure_derivatives'] = TRUE;
$config['image.settings']['suppress_itok_output'] = TRUE;

/**
 * Optimized assets path:
 *
 * A local file system path where optimized assets will be stored. This directory
 * must exist and be writable by Drupal. This directory must be relative to
 * the Drupal installation directory and be accessible over the web.
 */
$settings['file_assets_path'] = 'sites/default/files';

/**
 * CMS Build settings.
 *
 * These are settings to trigger a static file, front-end WEB build job.
 * DevOps engineers will need to get and set the getenv() ENV variables below
 * from va.gov-cms-devops Ansible Vault for testing.
 * @see /README.md for details
 */
$settings['jenkins_build_env'] = getenv('CMS_ENVIRONMENT_TYPE') ?: FALSE;
$settings['jenkins_build_job_host'] = 'http://jenkins.vfs.va.gov';
// Authorized to the Jenkins API via GitHub login.
$settings['va_cms_bot_github_username'] = 'va-cms-bot';
$settings['va_cms_bot_github_auth_token'] = getenv('CMS_GITHUB_VA_CMS_BOT_TOKEN') ?: FALSE;

// Environment settings
$settings['va_gov_composer_home'] = getenv('COMPOSER_HOME');
$settings['va_gov_path_to_composer'] = '/usr/local/bin/composer';
// The default project root locations. These settings are currently only used on Tugboat and local environments.
$app_root_base = getenv('DDEV_APPROOT') ?: getenv('TUGBOAT_ROOT') ?: getenv('RUNNER_TEMP') ?: getenv('EKS_APP_ROOT') ?: '/var/www/cms';
$settings['va_gov_app_root'] = $app_root_base;
$settings['va_gov_web_root'] = $app_root_base . '/web';
$settings['va_gov_vets_website_root'] = $app_root_base . '/docroot/vendor/va-gov/vets-website';
$settings['va_gov_next_build_root'] = getenv('TUGBOAT_ROOT') ? getenv('TUGBOAT_ROOT') . '/var/www/cms/next' : '/var/www/cms/next';
$settings['va_gov_next_build_vets_website_root'] = getenv('TUGBOAT_ROOT') ? getenv('TUGBOAT_ROOT') . '/var/www/cms/vets-website' : '/var/www/cms/vets-website';

// Defaults (should only be local that doesn't set these), default to dev for config_split
$config['config_split.config_split.dev']['status'] = TRUE;
$config['config_split.config_split.stg']['status'] = FALSE;
$config['config_split.config_split.prod']['status'] = FALSE;
$config['system.performance']['cache']['page']['use_internal'] = FALSE;
$config['system.performance']['css']['preprocess'] = FALSE;
$config['system.performance']['css']['gzip'] = FALSE;
$config['system.performance']['js']['preprocess'] = FALSE;
$config['system.performance']['js']['gzip'] = FALSE;
$config['system.performance']['response']['gzip'] = FALSE;
$config['views.settings']['ui']['show']['sql_query']['enabled'] = TRUE;
$config['views.settings']['ui']['show']['performance_statistics'] = TRUE;
$config['system.logging']['error_level'] = 'all';
$config['environment_indicator.indicator']['bg_color'] = '#05F901';
$config['environment_indicator.indicator']['fg_color'] = '#000000';
$config['environment_indicator.indicator']['name'] = 'Local';

$settings['config_sync_directory'] = '../config/sync';

// Setting the directory for default content sync
$settings['default_content_deploy_content_directory'] = '../content';

$env_type = getenv('CMS_ENVIRONMENT_TYPE') ?: 'ci';

/**
 * Environment discovery service settings, vended by the Environment Discovery
 * service (`va_gov.environment_discovery`).
 */
$settings['va_gov_environment'] = [
  'environment_raw' => $env_type,
  'is_cms_test' => getenv('CMS_APP_NAME') ?? '' === 'cms-test',
];

$config['govdelivery_bulletins.settings']['govdelivery_endpoint'] = getenv('CMS_GOVDELIVERY_ENDPOINT') ?: FALSE;
$config['govdelivery_bulletins.settings']['govdelivery_username'] = getenv('CMS_GOVDELIVERY_USERNAME') ?: FALSE;
$config['govdelivery_bulletins.settings']['govdelivery_password'] = getenv('CMS_GOVDELIVERY_PASSWORD') ?: FALSE;
// Configuration for Mapbox
$config['geocoder.geocoder_provider.mapbox']['configuration']['accessToken'] = getenv('MAPBOX_TOKEN_CMS');

// Set migration settings from environment variables.
$facility_api_urls = [
  getenv('CMS_VAGOV_API_URL') . '/services/va_facilities/v1/facilities?per_page=1000',
  getenv('CMS_VAGOV_API_URL') . '/services/va_facilities/v1/facilities?per_page=1000&page=2',
  getenv('CMS_VAGOV_API_URL') . '/services/va_facilities/v1/facilities?per_page=1000&page=3',
];

$facility_api_key = getenv('CMS_VAGOV_API_KEY');
$facility_migrations = [
  'va_node_health_care_local_facility',
  'va_node_facility_vba',
  'va_node_facility_nca',
  'va_node_facility_vet_centers',
  'va_node_facility_vet_centers_mvc',
  'va_node_facility_vet_centers_os',
  //Add other migrations here
];

// Use classic node migrations instead of complete.
// cf. https://www.drupal.org/node/3105503
$settings['migrate_node_migrate_type_classic'] = TRUE;
foreach ($facility_migrations as $facility_migration) {
  $config["migrate_plus.migration.{$facility_migration}"]['source']['urls'] = $facility_api_urls;
  $config["migrate_plus.migration.{$facility_migration}"]['source']['headers']['apikey'] = $facility_api_key;
}

// HTTP client settings
$settings['http_client_config']['timeout'] = 60;

// Variables for post_api.
$settings['post_api_endpoint_host'] = getenv('CMS_VAGOV_API_URL') ?: FALSE;
$settings['post_api_apikey'] = getenv('CMS_VAGOV_API_KEY') ?: FALSE;

// AWS Module settings
$settings['aws_access_key_id'] = getenv('AWS_ACCESS_KEY_ID') ?: FALSE;
$settings['aws_secret_access_key'] = getenv('AWS_SECRET_ACCESS_KEY') ?: FALSE;

// Slack Webhook URL for csm-notifications channel.
$settings['slack_webhook_url'] = getenv('CMS_VAGOV_SLACK_WEBHOOK_URL') ?: FALSE;
$config['slack.settings']['slack_webhook_url'] = $settings['slack_webhook_url'];

// Settings supporting broken link report import. Off by default.
$settings['broken_link_report_import_enabled'] = FALSE;
// Default prod location, overrideable by env var.
$settings['broken_link_report_location'] = getenv('CONTENT_RELEASE_BROKEN_LINK_REPORT') ?: 'https://vetsgov-website-builds-s3-upload.s3-us-gov-west-1.amazonaws.com/broken-link-reports/vagovprod-broken-links.json';

// Hide deprecation warnings during transition to PHP 8.1.
$error_reporting = (int) ini_get('error_reporting');
ini_set('error_reporting', $error_reporting & ~E_DEPRECATED & ~E_USER_DEPRECATED);

$settings_files = [
  // Environment specific settings
  __DIR__ . '/settings/settings.' . $env_type . '.php',
  // Fast 404 settings
  __DIR__ . '/settings/settings.fast_404.php',
  // Ansible moves this file into place during deploy, so if it is present we are in deploy mode.
  __DIR__ . '/settings/settings.deploy.active.php',
  // Local overrides
  __DIR__ . '/settings.local.php',
];

foreach ($settings_files as $file) {
  if (file_exists($file)) {
    include $file;
  }
}

/**
 * Preserve control characters (e.g. newlines) in text passed to syslog.
 * @see https://bugs.php.net/bug.php?id=77913
 * TL;DR: Ensure each watchdog entry becomes only a single line in syslog.
 */
ini_set('syslog.filter', 'raw');

// The VA_GOV_IN_DEPLOY_MODE is set in settings.deploy.active.php.
// The file is copied from settings.deploy.inactive.php by Ansible during deploys.
if (!empty($GLOBALS['request']) &&
  is_a($GLOBALS['request'], \Symfony\Component\HttpFoundation\Request::class) &&
  !empty(getenv('VA_GOV_IN_DEPLOY_MODE'))) {

  $deploy_service = \Drupal\va_gov_backend\Deploy\DeployService::create();
  $deploy_service->run($GLOBALS['request'], $app_root, $site_path);
}

// Because Jenkins can't resolve e.g. prod.cms.va.gov DNS, and only the
// internal*elb addresses. And sometimes the host would return data with
// "http://default/..." hostnames for files so we set the host here and pass it
// to the `file_public_base_url` setting to fix that.
if (!empty($webhost_on_cli)) {
  if (PHP_SAPI === 'cli') {
    // This is running from drush so set the webhost.
    // Var $webhost_on_cli is set in <settings.<environment>.php.
    $webhost = $webhost_on_cli;
  }
  else {
    // This is running from an HTTP request.
    $webhost = "{$_SERVER['REQUEST_SCHEME']}://{$_SERVER['HTTP_HOST']}";
  }

  $settings['file_public_base_url'] = "{$webhost}/sites/default/files";
}

// Look for an incoming request header to indicate we should use the public
// asset S3 location for file rather than the Drupal-internal location.
if (!empty($public_asset_s3_base_url)) {
  $headerArray = function_exists('apache_request_headers') ? apache_request_headers() : [];
  $targetHeader = 'File-Public-Base-Url-Check';
  foreach ($headerArray as $header => $value) {
    if (strtolower($header) == strtolower($targetHeader) && $value === 'true') {
      // Point the file base url to the public asset S3 bucket.
      $settings['file_public_base_url'] = $public_asset_s3_base_url;
    }
  }
 }

// Monolog
$settings['container_yamls'][] = __DIR__ . '/services/services.monolog.yml';

// Memcache-specific settings
if ((extension_loaded('memcache') || extension_loaded('memcached')) && !empty($settings['memcache']['servers'])) {
  $settings['cache']['default'] = 'cache.backend.memcache';
  $settings['memcache']['bins'] = [
    'default' => 'default',
  ];
  $settings['container_yamls'][] = __DIR__ . '/services/services.memcache.yml';
  $settings['memcache']['persistent'] = 'drupal';
}

// Environment specific services container.
$env_services_path = __DIR__ . "/services/services.$env_type.yml";
if (file_exists($env_services_path)) {
  $settings['container_yamls'][] = $env_services_path;
}

// Global override for setting the session transaction isolation level.
// This is intended to prevent deadlocks in the course of normal operation.
// @see https://www.drupal.org/project/drupal/issues/2733675
$databases['default']['default']['init_commands']['isolation_level'] = 'SET SESSION TRANSACTION ISOLATION LEVEL READ COMMITTED';


$settings['state_cache'] = TRUE;

// Disable the use of insecure uploads.
$config['system.file']['allow_insecure_uploads'] = FALSE;

$settings['session_write_interval'] = 30;
